README
ID3-FromHierarchy version 0.05
==============================

INSTALLATION
------------

To install this module type the following:

``` bash
   $ perl Makefile.PL
   $ make
   $ make test
   $ make install
```

DEPENDENCIES
------------

This module requires these other modules and libraries:

  - perl 5.10.0
  - Readonly

COPYRIGHT AND LICENCE
---------------------

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension ID3::FromHierarchy.

0.05  Sun Dec 28 13:29:05 JST 2014
    - Change from OO style to function style
0.04  Sat Nov 12 16:54:24 2011
    - Add PREREQ module.
0.03  Wed Fri 16 00:42:00 2011
    - Remove Mouse.
    - Re-factoring.
0.02  Thu Nov 29 16:54:40 2009
    - Also ok if music-file is named <track>-<title>.
0.01  Wed Dec  2 17:17:52 2009
    - original version; created by h2xs 1.23 with options
        -Xan ID3::FromHierarchy



eg/change.pl
#!/usr/bin/perl
use 5.10.0;
use utf8;
use strict;
use warnings;
use open qw( :utf8 :std );
use Readonly;
use Encode qw( decode );
use Encode::UTF8Mac;
use Path::Class qw( dir );
use List::MoreUtils qw( true );
use MP3::Tag;
use ID3::FromHierarchy;

Readonly my %FRAME => (
    genre  => "TCON",
    artist => "TPE1",
    album  => "TALB",
    track  => "TRCK",
    title  => "TIT2",
);

MP3::Tag->config( write_v24 => 1 );

my $root = dir( "./" ); # game/ODIN_SPHERE/valkyrie_introduction.mp3
my @target_genres = qw( game );

foreach my $genre ( grep { $_->is_dir } $root->children ) {
    next
        unless true { $_ eq $genre ->relative( $root ) } @target_genres;

    say "### ", decode( "utf-8-mac", $genre ->relative( $root ) );

    foreach my $artist ( grep { $_->is_dir } $genre->children ) {
        say "--- ", decode( "utf-8-mac", $artist->relative( $genre ) );

        foreach my $mp3 ( grep { !$_->is_dir } $artist->children ) {
            say "--- --- ", decode( "utf-8-mac", $mp3->relative( $artist ) );

            my %tag = eval { ID3::FromHierarchy::get_tag_from_filename( $mp3->relative( $root ) ) };

            if ( my $e = $@ ) {
                warn $e
                    and next;
            }

            print map { "--- --- --- $_: $tag{$_}\n" } grep { defined $tag{ $_ } } keys %tag;

            update_id3( $mp3, %tag );
        }

        foreach my $album ( grep { $_->is_dir } $artist->children ) {
            say "--- --- ", decode( "utf-8-mac", $album->relative( $artist ) );

            foreach my $mp3 ( grep { !$_->is_dir } $album->children ) {
                say "--- --- --- ", decode( "utf-8-mac", $mp3->relative( $album ) );

                my %tag = eval { ID3::FromHierarchy::get_tag_from_filename( $mp3->relative( $root ) ) };

                if ( my $e = $@ ) {
                    warn $e
                        and next;
                }

                print map { "--- --- --- --- $_: $tag{$_}\n" } grep { defined $tag{ $_ } } keys %tag;

                update_id3( $mp3, %tag );
            }
        }
    }
}

exit;

sub update_id3 {
    my( $file, %tag ) = @_;
    my $mp3 = MP3::Tag->new( $file );

    $mp3->{ID3v1}->remove_tag
        if exists $mp3->{ID3v1};
    $mp3->{ID3v2}->remove_tag
        if exists $mp3->{ID3v2};

    $mp3->new_tag( "ID3v2" );

    foreach my $key ( keys %tag ) {
        $mp3->{ID3v2}->add_frame( $FRAME{ $key }, $tag { $key } );
    }

    $mp3->{ID3v2}->write_tag;
    $mp3->close;

    return;
}



lib/ID3/FromHierarchy.pm
package ID3::FromHierarchy;
use utf8;
use strict;
use warnings;
use Readonly;
use Carp qw( carp croak );
use File::Basename qw( basename );
use File::Spec::Functions qw( splitdir catfile );

our $VERSION = "0.05";

Readonly my $IS_MP3_RE    => qr{ [.] mp3 }i;
Readonly my $TRACK_PARSER => qr{\A (\d{1,3}) (?:[ ][-][ ] | [ ]) (.*) }msx;

sub _make_genre_artist_title {
    my @paths = @_;
    my %tag = (
        title   => basename( $paths[2], $IS_MP3_RE ),
        track   => undef,
        artist  => $paths[1],
        album   => undef,
        genre   => $paths[0],
    );
    return %tag;
}

sub _make_genre_artist_album_title {
    my @paths = @_;
    my %tag = (
        title   => basename( $paths[3], $IS_MP3_RE ),
        track   => undef,
        artist  => $paths[1],
        album   => $paths[2],
        genre   => $paths[0],
    );
    return %tag;
}

sub _get_tag_from_paths {
    return &_make_genre_artist_title
        if @_ == 3;
    return &_make_genre_artist_album_title
        if @_ == 4;
    croak "Paths should be 3, or 4.";
}

sub _split_title_into_track_and_path {
    my %tag = @_;
    if ( $tag{title} =~ $TRACK_PARSER ) {
        @tag{ qw( track title ) } = ( $1, $2 );
    }
    return @tag{ qw( track title ) };
}

sub _split_album_into_number_and_name {
    my %tag = @_;
    if ( $tag{album} =~ $TRACK_PARSER ) {
        @tag{ qw( disk_number album ) } = ( $1, $2 );
    }
    return @tag{ qw( disk_number album ) };
}

sub get_tag_from_filename {
    my $filename = shift
        or croak "filename required.";

    my @paths = splitdir( $filename );
    croak "The depth of target is not neither 3 nor 4.  This requires relative path."
        if @paths != 3 && @paths != 4;
    my $has_album = @paths == 4;

    my %tag = _get_tag_from_paths( @paths );
    @tag{ qw( track title ) } = _split_title_into_track_and_path( %tag );

    @tag{ qw( disk_number album ) } = _split_album_into_number_and_name( %tag )
        if $has_album;

    return %tag;
}

1;

__END__
=head1 NAME

ID3::FromHierarchy - Parses ID3 tag from path hierarchy.

=head1 SYNOPSIS

  use ID3::FromHierarchy;
  my $file = "path/of/mp3filename";
  my %tag  = ID3::FromHierarchy::get_tag_from_filename( $file );

=head1 DESCRIPTION

Make tag data from filename.
Note that this module does not read the file to get ID3 tag.

=head1 FUNCTIONS

=over

=item get_tag_from_filename( $filename )

Returns ID3 tag daat from filename which can be:

  - <genre>/<artist>/<album_number> - <album_name>/<track> - <title>.mp3
  - <genre>/<artist>/<album_number> - <album_name>/<title>.mp3
  - <genre>/<artist>/<album>/<title>.mp3
  - <genre>/<artist>/<album>/<track> - <title>.mp3
  - <genre>/<artist>/<track> - <title>.mp3
  - <genre>/<artist>/<title>.mp3

=back

=head1 SEE ALSO

=over

=item MP3::Tag

=back

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/ID3/FromHierarchy/GenreArtistTitle.pm
package ID3::FromHierarchy::GenreArtistTitle;

use 5.010000;
use utf8;
use Modern::Perl;
use Carp qw( carp croak );
use Readonly;
use base "Class::Accessor";
use File::Basename qw( basename );
use File::Spec::Functions qw( splitdir );

__PACKAGE__->mk_accessors( qw(
    tag
) );

our $VERSION = '0.03';

Readonly my @SUFFIXES => qw( .mp3 .MP3 );

sub parse {
    my $self   = shift;
    my $target = shift
        or croak "Target required.";

    my @pathes = splitdir( $target );
    croak "Could not parse. Its depth is too shallow."
        unless @pathes > 2;
    croak "Could not parse: Its depth is too deep."
        unless @pathes < 4;

    my %tag = (
        title   => basename( $pathes[2], @SUFFIXES ),
        track   => undef,
        artist  => $pathes[1],
        album   => undef,
        genre   => $pathes[0],
    );

    # Verify title, and set track.
    given ( $tag{title} ) {
        when ( m{\A (\d{2}) [ ] [-] [ ] (.*) \z}msx ) {
            $tag{track} = $1;
            $tag{title} = $2;
        }
        when ( m{\A (\d{2}) [-] (.*) \z}msx )  {
            $tag{track} = $1;
            $tag{title} = $2;
        }
        when ( m{\A (\d{2}) [ ] (.*) \z}msx ) {
            $tag{track} = $1;
            $tag{title} = $2;
        }
    }

    $self->tag( \%tag );

    return %{ $self->tag };
}


1;
__END__
=head1 NAME

ID3::FromHierarchy::GenreArtistTitle - Perl extension for ID3::FromHierarchy

=head1 SYNOPSIS

  use ID3::FromHierarchy::GenreArtistTitle;
  use File::Spec;
  my $file = File::Spec->catfile( "genre", "artist", "title" );
  my %tag  = eval {
      ID3::FromHierarchy::GenreArtistTitle->new->parse( $file )
  };
  die $@
      if ( $@ );
  print map { "$_: $tag{$_}\n" } keys %tag;

=head1 DESCRIPTION

In this module expects the file system has hierarchy like genre - artist - music file.
According to this expection, parses ID3 tag, and returns tag.

=head1 ATTRIBUTES

=over

=item tag

=back

=head1 METHODS

=over

=item parse( $filename )

Gets filename, and returns ID3 tag.

=back

=head1 SEE ALSO

=over

=item ID3::FromHierarchy

=item ID3::FromHierarchy::GenreArtistAlbumTitle

=back

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/ID3/FromHierarchy/GenreArtistAlbumTitle.pm
package ID3::FromHierarchy::GenreArtistAlbumTitle;

use 5.010000;
use utf8;
use Modern::Perl;
use Carp qw( carp croak );
use Readonly;
use base "Class::Accessor";
use File::Basename qw( basename );
use File::Spec::Functions qw( splitdir );

__PACKAGE__->mk_accessors( qw(
    tag
) );

our $VERSION = '0.03';

my @SUFFIXES = qw( .mp3 .MP3 );

sub parse {
    my $self   = shift;
    my $target = shift
        or croak "Target required.";

    my @pathes = splitdir( $target );
    croak "Could not parse. Its depth is too shallow."
        unless @pathes > 3;
    croak "Could not parse: Its depth is too deep."
        unless @pathes < 5;

    my %tag = (
        title   => basename( $pathes[3], @SUFFIXES ),
        track   => undef,
        artist  => $pathes[1],
        album   => $pathes[2],
        genre   => $pathes[0],
    );

    # Velify title, and set track.
    given ( $tag{title} ) {
        when ( m{\A (\d{2}) [ ] [-] [ ] (.*) \z}msx ) {
            $tag{track} = $1;
            $tag{title} = $2;
        }
        when ( m{\A (\d{2}) [-] (.*) \z}msx ) {
            $tag{track} = $1;
            $tag{title} = $2;
        }
        when ( m{\A (\d{2}) [ ] (.*) \z}msx ) {
            $tag{track} = $1;
            $tag{title} = $2;
        }
    }

    # Trim disc number from album.
    if ( $tag{album} =~ m{\A ([sab]? \d{2} [ ] [-] [ ]) (.*) \z}msx ) {
        $tag{album} = $2;
    }
    elsif ( $tag{album} =~ m{\A ([sab]? \d{2} [ ]) (.*) \z}msx ) {
        $tag{album} = $2;
    }

    $self->tag( \%tag );

    return %{ $self->tag };
}


1;
__END__
=head1 NAME

ID3::FromHierarchy::GenreArtistAlbumTitle - Perl extension for ID3::FromHierarchy

=head1 SYNOPSIS

  use ID3::FromHierarchy::GenreArtistAlbumTitle;
  use File::Spec;
  my $file = File::Spec->catfile( "genre", "artist", "title" );
  my %tag  = eval {
      ID3::FromHierarchy::GenreArtistAlbumTitle->new->parse( $file )
  };
  die $@
      if ( $@ );
  print map { "$_: $tag{$_}\n" } keys %tag;

=head1 DESCRIPTION

In this module expects the file system has hierarchy like
genre - artist - album - music file.
According to this expection, parses ID3 tag, and returns tag.

=head1 ATTRIBUTES

=over

=item tag

=back

=head1 METHODS

=over

=item parse( $filename )

Gets filename, and returns ID3 tag.

=back

=head1 SEE ALSO

=over

=item ID3::FromHierarchy

=item ID3::FromHierarchy::GenreArtistTitle

=back

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut

